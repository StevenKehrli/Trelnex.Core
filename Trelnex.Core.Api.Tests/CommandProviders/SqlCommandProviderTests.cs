using Azure.Core;
using Azure.Identity;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Trelnex.Core.Api.CommandProviders;
using Trelnex.Core.Api.Serilog;
using Trelnex.Core.Data;
using Trelnex.Core.Data.Tests.CommandProviders;

namespace Trelnex.Core.Api.Tests.CommandProviders;

[Ignore("Requires a SQL server.")]
public class SqlCommandProviderTests : CommandProviderTests
{
    private readonly string _scope = "https://database.windows.net/.default";

    private TokenCredential _tokenCredential = null!;
    private string _connectionString = null!;
    private string _tableName = null!;

    [OneTimeSetUp]
    public void TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the service collection
        var services = new ServiceCollection();

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var dataSource = configuration
            .GetSection("SqlCommandProviders:DataSource")
            .Value;

        var initialCatalog = configuration
            .GetSection("SqlCommandProviders:InitialCatalog")
            .Value;

        _tableName = configuration
            .GetSection("SqlCommandProviders:Tables:0:TableName")
            .Value!;

        var scsBuilder = new SqlConnectionStringBuilder()
        {
            DataSource = dataSource,
            InitialCatalog = initialCatalog,
            Encrypt = true,
        };

        _connectionString = scsBuilder.ConnectionString;

        // create the command provider
        _tokenCredential = new DefaultAzureCredential();

        var bootstrapLogger = services.AddSerilog(
            configuration,
            "Trelnex.Integration.Tests");

        services.AddSqlCommandProviders(
            configuration,
            bootstrapLogger,
            options => options.Add<ITestItem, TestItem>(
                typeName: "test-item",
                validator: TestItem.Validator,
                commandOperations: CommandOperations.All));

        var serviceProvider = services.BuildServiceProvider();

        // get the command provider
        _commandProvider = serviceProvider.GetRequiredService<ICommandProvider<ITestItem>>();
    }

    [TearDown]
    public void Cleanup()
    {
        // This method is called after each test has run.
        using var sqlConnection = new SqlConnection(_connectionString);

        var tokenRequestContext = new TokenRequestContext([ _scope ]);
        sqlConnection.AccessToken = _tokenCredential.GetToken(tokenRequestContext, default).Token;

        sqlConnection.Open();

        var cmdText = $"DELETE FROM [{_tableName}-events]; DELETE FROM [{_tableName}];";
        var sqlCommand = new SqlCommand(cmdText, sqlConnection);

        sqlCommand.ExecuteNonQuery();
    }
}
