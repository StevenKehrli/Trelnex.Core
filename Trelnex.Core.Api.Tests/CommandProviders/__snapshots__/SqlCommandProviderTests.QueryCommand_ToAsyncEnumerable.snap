[
  {
    "Item": {
      "Id": "3fca6d8a-75c1-491a-9178-90343551364a",
      "PartitionKey": "81dc4acd-dcbe-4d5f-a36f-21a35f158b2c",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-04T01:26:26.710549Z",
      "UpdatedDate": "2025-01-04T01:26:26.710549Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "98050F7C-F299-48A6-98F4-7904CB81ADEF",
      "Message": "Message #1"
    }
  },
  {
    "Item": {
      "Id": "648de92a-b7e8-41c5-a5d2-bdf0cc65d67c",
      "PartitionKey": "e36f287e-188d-4a74-9db7-dab74282b5dd",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-04T01:26:27.58377Z",
      "UpdatedDate": "2025-01-04T01:26:27.58377Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "2DDF539F-52A1-49E5-A7E7-8283D68556EA",
      "Message": "Message #2"
    }
  }
]
